link/uprobe_test.go

```diff
@@ -88,6 +88,20 @@ func TestUprobeExtWithOpts(t *testing.T) {
 	defer up.Close()
 }
 
+func TestUprobeWithPid(t *testing.T) {
+	prog, err := ebpf.NewProgram(&kprobeSpec)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer prog.Close()
+
+	up, err := bashEx.Uprobe(bashSym, prog, &UprobeOptions{Pid: -1})
```

Comment:
with pid `-1` this is guaranteed to work, even if it's technically useless. The alternative is to obtain a random pid from the current list of processes (eg. by reading `/proc`)

Reply:
Oh, does the kernel reject probes on invalid PIDs?

What about pid1? It's pretty unlikely to `exec()` into Bash. :wink: If the kernel accepts the probe on pid1, and there's no eBPF program invocation when executing bash, we (partially) have our negative test coverage.

Reply:
yes, this is the error:

```
=== RUN   TestUprobeWithPid
    uprobe_test.go:100: creating perf_uprobe PMU: opening perf event: no such process
```

I can use pid 1 👍🏻

Reply:
Oh, even better! Then we could try to attach a probe to `os.Getpid()` for the happy path, and one to a nonsensical PID and expect a `unix.ESRCH`. `math.MaxInt32` seems to elicit an ESRCH on my machine.

Reply:
Ok for the happy path. Not 100% convinced on the other one since *in theory* any pid which is a positive `int` could exist; what about using `-2`? It is still an `int` but we are sure that no process with that pid exist

Reply:
> in theory any pid which is a positive `int` could exist

The set of possible PIDs is bounded by `/proc/sys/kernel/pid_max`, which evaluates to `4194304` on my machine. MaxInt32 is a few orders of magnitude larger, so seems good enough for our testing. Of course, if a negative PID has the same effect, that would be cleaner :+1:

Reply:
yes, it has the same effect (but I didn't knew about `pid_max` 😅 , so, I can use either)

Reply:
Let's go for -2 then :+1: